agent/xds/clusters.go

```diff
@@ -133,32 +133,56 @@ func (s *Server) clustersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsho
 		if dc == cfgSnap.Datacenter {
 			continue // skip local
 		}
-		clusterName := connect.DatacenterSNI(dc, cfgSnap.Roots.TrustDomain)
+		useEDS := true
+		if len(cfgSnap.MeshGateway.HostnameDatacenters[dc]) > 0 {
+			useEDS = false
+		}
 
-		cluster, err := s.makeGatewayCluster(cfgSnap, clusterName)
+		cluster := &envoy.Cluster{
+			Name: connect.DatacenterSNI(dc, cfgSnap.Roots.TrustDomain),
+		}
+		err := s.populateGatewayCluster(cfgSnap, cluster, 0, useEDS)
 		if err != nil {
 			return nil, err
 		}
+		if !useEDS {
+			injectHostnameEndpoints(cluster, cfgSnap.MeshGateway.HostnameDatacenters[dc], dc != cfgSnap.Datacenter, false)
+		}
+
 		clusters = append(clusters, cluster)
 	}
 
 	if cfgSnap.ServiceMeta[structs.MetaWANFederationKey] == "1" && cfgSnap.ServerSNIFn != nil {
 		// Add all of the remote wildcard datacenter mappings for servers.
 		for _, dc := range datacenters {
-			clusterName := cfgSnap.ServerSNIFn(dc, "")
+			if dc == cfgSnap.Datacenter {
+				continue // skip local
+			}
+			useEDS := true
+			if len(cfgSnap.MeshGateway.HostnameDatacenters[dc]) > 0 {
+				useEDS = false
+			}
 
-			cluster, err := s.makeGatewayCluster(cfgSnap, clusterName)
+			cluster := &envoy.Cluster{
+				Name: cfgSnap.ServerSNIFn(dc, ""),
+			}
+			err := s.populateGatewayCluster(cfgSnap, cluster, 0, useEDS)
 			if err != nil {
 				return nil, err
 			}
+			if !useEDS {
+				injectHostnameEndpoints(cluster, cfgSnap.MeshGateway.HostnameDatacenters[dc], dc != cfgSnap.Datacenter, false)
+			}
+
```

Comment:
This block and the one above it seem to be repeated. Would it be possible to move all of this logic inside `makeGatewayCluster` and kept the old code here? We already pass `cfgSnap` so it should have all the info it needs to make do the same logic unless I'm missing something.

Reply:
Refactored this